internal/semaphore/semaphore_test.go

```diff
@@ -0,0 +1,93 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package semaphore
+
+import (
+	"sync"
+	"testing"
+	"github.com/stretchr/testify/assert"
+)
+
+func checkLimit(t *testing.T, sem Semaphore, expected int) {
+	limit := sem.GetLimit()
+	assert.Equal(t, expected, limit)
+}
+
+func checkCount(t *testing.T, sem Semaphore, expected int) {
+	count := sem.GetCount()
+	assert.Equal(t, expected, count)
+}
+
+func checkLimitAndCount(t *testing.T, sem Semaphore, expectedLimit, expectedCount int) {
+	checkLimit(t, sem, expectedLimit)
+	checkCount(t, sem, expectedCount)
+}
+
+func TestNewSemaphore_NegativeLimit(t *testing.T) {
+	assert.Panics(t, func() { _ = NewSemaphore(-1) })
+}
+
+func TestSemaphore_AcquireRelease(t *testing.T) {
+	sem := NewSemaphore(2)
+	checkLimitAndCount(t, sem, 2, 0)
+
+	sem.Acquire()
+	checkLimitAndCount(t, sem, 2, 1)
+
+	sem.Acquire()
+	checkLimitAndCount(t, sem, 2, 2)
+
+	sem.Release()
+	checkLimitAndCount(t, sem, 2, 1)
+
+	sem.Release()
+	checkLimitAndCount(t, sem, 2, 0)
+}
+
+func TestSemaphore_ReleaseMoreThanAcquire(t *testing.T) {
+	sem := NewSemaphore(1)
+	assert.Panics(t, func() { sem.Release() })
+}
+
+func TestSemaphore_AcquireReleaseUnderLimit(t *testing.T) {
+	sem := NewSemaphore(100)
+
+	c := make(chan struct{})
+	wg := sync.WaitGroup{}
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			<-c
+			sem.Acquire()
+			sem.Release()
+			wg.Done()
+		}()
+	}
+
+	close(c) // start
+	wg.Wait()
+
+	checkLimitAndCount(t, sem, 100, 0)
+}
+
+func TestSemaphore_AcquireReleaseOverLimit(t *testing.T) {
+	sem := NewSemaphore(10)
+
+	c := make(chan struct{})
+	wg := sync.WaitGroup{}
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		go func() {
+			<-c
+			sem.Acquire()
+			sem.Release()
+			wg.Done()
+		}()
+	}
+
+	close(c) // start
+	wg.Wait()
+
+	checkLimitAndCount(t, sem, 10, 0)
+}
```

Comment:
Not a blocker, but these are the same test with different parameters. This could be consolidated to a single test that has a set of test parameters that get executed with sub-tests via `t.Run()`

Reply:
Yes. Thought about `table driven test`; however, not sure if the `table driven test` are independent to each other when its come to go channel.